### PR TITLE
Allow iOS execute protocol to return value

### PR DESCRIPTION
**Pull request by cdierkens**
## Proposed changes

The previous pull request cut a little too deep with the IOS change. This PR takes the logic tree back to a single path, and returns the value from the execute command.

Without the 'return' statement, a call to an JSONWire implementing API will return NULL for the value. See: http://bit.ly/2g838LM 

## Types of changes

What types of changes does your code introduce to WebdriverIO?
- [x ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x ] I have read the [CONTRIBUTING](/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: lib/protocol/execute.js [hunk 1]
```diff
@@ -48,16 +48,17 @@ let execute = function (...args) {
         throw new ProtocolError('number or type of arguments don\'t agree with execute protocol command')
     }
 
-    if (this.isIOS) {
-        const strArgs = getArgs(args)
-        script = typeof script === 'string' ? script : `(${script}).apply(null, ${strArgs})`
```
Comment (cdierkens) [inline lib/protocol/execute.js]: `(${script}).apply(null, ${strArgs})` ->  `return (${script}).apply(null, ${strArgs})`

**Comment by christian-bromann** [comment]
Two things:
1. Can we have a test for this?
2. Can you remove the merge commit?

**Comment by cdierkens** [comment]
Can you forward instructions on building the test .app file since it's not
included in the repo? I would need that in order to run grunt test:ios.

On Wed, Nov 16, 2016 at 1:05 PM, Christian Bromann <notifications@github.com

**Comment by christian-bromann** [comment]
@cdierkens you find the app here: http://webdriverjs.christian-bromann.com/ios.zip. You can also build it by calling `$ grunt cordovacli:build_ios` but make sure you install `grunt-cordovacli` and `cordova`. They are not in the dev dependencies because they don't pass the snyk test.

**Comment by christian-bromann** [comment]
ping @cdierkens

**Comment by cdierkens** [comment]
@christian-bromann I've updated the PR, but I'm having issues getting the tests to run.

```bash
grunt test:ios
Running "test:ios" (test) task

Running "connect:testpage" (connect) task
Started connect web server on http://localhost:8080

Running "mochaTest:ios" (mochaTest) task



=======================================================================================
Selenium 2.0 / webdriver protocol bindings implementation with helper commands in nodejs.
For a complete list of commands, visit http://webdriver.io/api.html.
=======================================================================================

Running job with following conf: { baseUrl: 'http://127.0.0.1:8080',
  testPage: 
   { start: 'http://127.0.0.1:8080/index.html',
     subPage: 'http://127.0.0.1:8080/two.html',
     gestureTest: 'http://127.0.0.1:8080/gestureTest.html',
     staleTest: 'http://127.0.0.1:8080/stale.html',
     title: 'WebdriverIO Testpage' },
  host: 'localhost',
  port: 4723,
  path: undefined,
  logLevel: 'verbose',
  waitforTimeout: 1000,
  desiredCapabilities: 
   { app: '/Users/cdierkens/webdriverio/test/site/platforms/ios/build/emulator/WebdriverIO Guinea Pig.app',
     appPackage: 'io.webdriverio.example',
     appActivity: '.MainActivity',
     platformName: 'iOS',
     platformVersion: '9.2',
     deviceName: 'iPhone 6',
     'appium-version': '1.4.16' } }
[13:52:07]  COMMAND	POST 	 "/wd/hub/session"
[13:52:07]  DATA		{"desiredCapabilities":{"javascriptEnabled":true,"locationContextEnabled":true,"handlesAlerts":true,"rotatable":true,"app":"/Users/cdierkens/webdriverio/test/site/platforms/ios/build/emulator/WebdriverIO Guinea Pig.app","appPackage":"io.webdriverio.example","appActivity":".MainActivity","platformName":"iOS","platformVersion":"9.2","deviceName":"iPhone 6","_":"2 more keys: [\"appium-version\",\"requestOrigins\"]"}}
  1) "before all" hook
[13:53:59]  COMMAND	GET 	 "/wd/hub/sessions"
[13:53:59]  DATA		{}
[13:53:59]  RESULT		[{"id":"1ba6b991-45be-4f33-b460-7e198b459788","capabilities":{"webStorageEnabled":false,"locationContextEnabled":true,"browserName":"","platform":"MAC","javascriptEnabled":true,"databaseEnabled":false ... (906 more bytes)
[13:53:59]  COMMAND	DELETE 	 "/wd/hub/session/1ba6b991-45be-4f33-b460-7e198b459788"
[13:53:59]  DATA		{}
  2) "after all" hook

  0 passing (3m)
  2 failing

  1)  "before all" hook:
     Error: A new session could not be created. Details: Problem getting session data for driver type IosDriver; does it implement 'get driverData'?
  

  2)  "after all" hook:
     Error: An unknown server-side error occurred while processing the command. Original error: Command 'killall Simulator' exited with code 1
  



Warning: Task "mochaTest:ios" failed. Use --force to continue.

Aborted due to warnings.
```

**Comment by christian-bromann** [comment]
@cdierkens I don't know if you need `appPackage` and `appAcitivity`. Check out this example: https://github.com/webdriverio/webdriverio/blob/master/examples/appium/appium-draws-its-logo.js where I use the test app.

**Comment by cdierkens** [comment]
@christian-bromann I don't have the liberty to spend time diagnosing why the CI/CD pipeline is failing. As it currently stands, all IOS webview contexts will remain broken, as no calls to execute will ever return a value. That was a result of a previous BREAKING PR. I did not change enough that the entire CI/CD pipeline should return RED, when it previously PASSED all standard tests. If you would like to make the test execution easier for me I would be happy to continue on, however, I do not have enough expertise in your specialized test execution for mobile devices for my time to be spent efficiently on fixing it.

**Comment by christian-bromann** [comment]
@cdierkens can you at least update your branch? I can take it from there then.

**Comment by cdierkens** [comment]
@christian-bromann What updates are you looking for?

**Comment by christian-bromann** [comment]
> What updates are you looking for?

Just rebase this branch on top of current master. Thanks!

**Comment by christian-bromann** [comment]
> Problem getting session data for driver type IosDriver; does it implement 'get driverData'?

@cdierkens I ran into the same issue, seems to be a race condition of some sort. Rerunning the test worked for me.

After some investigations it seems to be an issue in appiums ios driver. It doesn't seem to execute the script within a function body as it states in [the protocol](https://www.w3.org/TR/webdriver/#execute-script). I raised an issue here: https://github.com/appium/appium-ios-driver/issues/199 as well as removed the if statement in execute.js completely.

**Pull request closed without merging**